src/Uno.UWP/Globalization/NumberFormatting/NumeralSystemTranslatorHelper.cs

```diff
@@ -0,0 +1,3566 @@
+				case "th":
+					return "Latn";
+				case "th-th":
+					return "Latn";
+				case "bo":
+					return "Latn";
+				case "bo-in":
+					return "Latn";
+				case "bo-cn":
+					return "Latn";
+				case "tig":
+					return "Latn";
+				case "tig-er":
+					return "Latn";
+				case "ti":
+					return "Latn";
+				case "ti-er":
+					return "Latn";
+				case "ti-et":
+					return "Latn";
+				case "to":
+					return "Latn";
+				case "to-to":
+					return "Latn";
+				case "ts":
+					return "Latn";
+				case "ts-za":
+					return "Latn";
+				case "tr":
+					return "Latn";
+				case "tr-cy":
+					return "Latn";
+				case "tr-tr":
+					return "Latn";
+				case "tk":
+					return "Latn";
+				case "tk-tm":
+					return "Latn";
+				case "uk":
+					return "Latn";
+				case "uk-ua":
+					return "Latn";
+				case "hsb":
+					return "Latn";
+				case "hsb-de":
+					return "Latn";
+				case "ur":
+					return "Latn";
+				case "ur-in":
+					return "ArabExt";
+				case "ur-pk":
+					return "Latn";
+				case "ug":
+					return "Latn";
+				case "ug-cn":
+					return "Latn";
+				case "uz-arab":
+					return "ArabExt";
+				case "uz-arab-af":
+					return "ArabExt";
+				case "uz-cyrl":
+					return "Latn";
+				case "uz-cyrl-uz":
+					return "Latn";
+				case "uz":
+					return "Latn";
+				case "uz-latn":
+					return "Latn";
+				case "uz-latn-uz":
+					return "Latn";
+				case "vai":
+					return "Latn";
+				case "vai-vaii":
+					return "Latn";
+				case "vai-vaii-lr":
+					return "Latn";
+				case "vai-latn-lr":
+					return "Latn";
+				case "vai-latn":
+					return "Latn";
+				case "ca-es-valencia":
+					return "Latn";
+				case "ve":
+					return "Latn";
+				case "ve-za":
+					return "Latn";
+				case "vi":
+					return "Latn";
+				case "vi-vn":
+					return "Latn";
+				case "vo":
+					return "Latn";
+				case "vo-001":
+					return "Latn";
+				case "vun":
+					return "Latn";
+				case "vun-tz":
+					return "Latn";
+				case "wae":
+					return "Latn";
+				case "wae-ch":
+					return "Latn";
+				case "cy":
+					return "Latn";
+				case "cy-gb":
+					return "Latn";
+				case "wal":
+					return "Latn";
+				case "wal-et":
+					return "Latn";
+				case "wo":
+					return "Latn";
+				case "wo-sn":
+					return "Latn";
+				case "xh":
+					return "Latn";
+				case "xh-za":
+					return "Latn";
+				case "yav":
+					return "Latn";
+				case "yav-cm":
+					return "Latn";
+				case "ii":
+					return "Latn";
+				case "ii-cn":
+					return "Latn";
+				case "yi-001":
+					return "Latn";
+				case "yo":
+					return "Latn";
+				case "yo-bj":
+					return "Latn";
+				case "yo-ng":
+					return "Latn";
+				case "dje":
+					return "Latn";
+				case "dje-ne":
+					return "Latn";
+				case "zu":
+					return "Latn";
+				case "zu-za":
+					return "Latn";
+				default:
+					throw new ArgumentException("The parameter is incorrect.");
+			}
+		}
+
+		public static string GetResolvedLanguage(string languageTag)
```

Comment:
- `new Windows.Globalization.NumberFormatting.NumeralSystemTranslator(new[] { "zh-Hans-CN" })` doesn't throw in UWP. Looks like this code will throw?. (Note that `zh-Hans-CN` produces `ResolvedLanguage` = `zh-Hans-CN`, but `fr-TEST` will produce `en-US`.

- I think this logic is common between the `Language` class as well, and also many other globalization APIs. So we might in the future want to move it somewhere less-specific to `NumberFormatting`.

Reply:
I will check zh-Hans-CN thanks. Do you have complete list of language tags to prevent missing one?

I don't understand this part "but fr-TEST will produce en-US" what is fr-test?

Reply:
`fr-TEST` is just an invalid string I tried 😄

Reply:
> Do you have complete list of language tags to prevent missing one

I don't, I just knew about it since I tried to work on globalization-related fixes and got this test failing:

https://github.com/unoplatform/uno/blob/master/src/Uno.UI.RuntimeTests/Tests/Windows_Globalization/Given_ApplicationLanguages.cs#L21

Reply:
Some information I got from Microsoft:

---

```
The logic of language tags is based on the RFCs and other considerations.  However, because of geopolitical and other considerations, this space is not stable.  Tags can (and do) change between Windows versions. 

 

Since Windows doesn’t agree with other Windows versions, it isn’t practical to expect that another implementation would agree with either of those.

 

fr-Latn-CA has a “suppress-script” flag because fr-CA is sufficient as it’s always Latin.  Some other languages like az-Cyrl or az-Latn keep the script because the language may be written in multiple scripts.

 

Windows also sometimes choses only the language name when there’s typically no ambiguity over language.  Eg: there are lots of “versions” of English; en-CA, en-US, en-AU, etc… but only one version of Japanes:  ja.  The JP in ja-JP is therefore somewhat redundant.

 

Chinese is a little more confusing is different vendors have different expectations of the language tags for Chinese.  In particular, the Cantonese vs Mandarin differences.

 

Additionally, some tags are “aliased” because the standard names &/or the way that the industry handled those languages shifted over time.

 

Furthermore, any tag for which we have limited data may not be normalized in such a fashion.

 

Generally, the recommendation is to be flexible in what language tag(s) are accepted and used as they may change over time.

 

IETF language tag - Wikipedia

 

The “suppress script” tags are defined in IANA.  https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry

 

The tags for which the region is dropped is somewhat arbitrary and may change.  Frankly, for an alternate implementation, I’d probably be happy if the region wasn’t dropped.
```

---

```
Regarding the behavior of the Language constructor in Windows.Globalization:

 

> When we call `new Language("fr-Latn-CA")`, the language tag becomes "fr-CA" (the middle part dropped)
> However, when we call the ctor passing "zh-Hans-CN", nothing drops in LanguageTag


The reason that the script subtag (Latn, Hans) is sometimes dropped (or kept) depends on if the language subtag has a “Suppress-Script” defined by IANA in the Language Subtag Registry.

 

The IANA subtag registry is here: https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry

 

In the BCP47 spec / RFC5646 spec (https://www.rfc-editor.org/rfc/rfc5646.html) in section 4.1 says


The script subtag SHOULD NOT be used to form language tags unless
the script adds some distinguishing information to the tag.
 

So if the language has a suppress-script defined, then it is omitted from the overall tag.

 

For example, for French (fr), there is a suppress script defined of Latin/Latn:

 

%%

Type: language

Subtag: fr

Description: French

Added: 2005-10-16

Suppress-Script: Latn

 

This is why fr-Latn-CA ends up with fr-CA, as the Latn part is redundant.

 

The internal implementation has a partial copy of the IANA subtag registry, which is used to determine the suppress-script (if any).

 

 

> Moreover, calling it  with "az-Latn-AZ", the last part drops:


The dropping of the region part of the tag is unfortunately much less well defined.

AFAIK, there is no requirement for this in the BCP47 spec, though it does say in section 4.1 to:

 

Use as precise a tag as possible, but no more specific than is
justified.  Avoid using subtags that are not important for
distinguishing content in an application.
 

So for some tags, the regional part isn’t considered distinguishing, and it is omitted.

For example, ja-JP drops the region part JP, and is just “ja”, as the thinking was that there isn’t any distingusing data added by the region JP for the language Japanese.

 

There is no external data source for what is considered distingusing for the regoinal subtag, so you’d likely need to create a table to replicate it.
```

---


```
I’d like to reiterate that we pick up updated suppress-script data on a sort of ad-hoc basis.  Best practice would probably just be to use whatever’s current and recognize that it may differ from the behavior of any specific version of Windows.

 

And, for the dropping of the region, you could probably query a current Windows box with all of the registered IANA tags and start with that.  But as Jeff says, there’s no standard and that’s far more likely to change.
```

Reply:
Please note that information in the previous comment was an answer to `Windows.Globalization.Language` class. I'm posting assuming they follow the same behavior.

Reply:
Thanks for the information. It seems that the resolved language is your default windows language when the input language tag is "tag-InvalidSomething". I think [Language ](https://github.com/unoplatform/uno/blob/287f9785be664ca577775b1bb6b90fe68bbd26fb/src/Uno.UWP/Globalization/Language.cs#L9)class needs work because CultureInfo throws exception when input is "zh-cmn-Hans" but Language doesn't. I'm confused :)

Reply:
Yeah `Language` is currently buggy. It has been preventing me from fixing `PrimaryLanguageOverride` (https://github.com/unoplatform/uno/pull/6906)

Reply:
It seems that for best implementation RFCs should be read this needs time. Can we merge this and then complete the implementation in the future? Or we can mark the NumeralSystemTranslator constructor with parameter as not implemented. Users can set NumeralSystem property manually and call TranslateNumeral.